### ISSUE TITLE
New reference files needed for filter-dependent wander

### ISSUE THREAD

**Issue by stscijgbot** [spacetelescope/jwst#4555]
Issue [JP-1290](https://jira.stsci.edu/browse/JP-1290) was created by Alicia Canipe:

New FILTER_OFFSET reference files are needed due to address filter-dependent wander mentioned in https://jira.stsci.edu/browse/JSOCINT-348.

Corresponding Jira tickets for to define the calibration pipeline updates, datamodel/reference file format, and CRDS delivery are attached to this epic.

**Comment by stscijgbot** [comment]
Comment by Steven Crawford: [~acanipe] I think we should discuss the implementation here before proceeding.   

From my understanding of what is needed along with the design of the calibration software, our recommendation for a solution would be that each filter should have its own DISTORTION reference file.    

The introduction of the new filter offsets file would be additional code to support that would introduce a redundancy with the existing functionality already in CRDS and the calibration software.     

This solution would end up in a small amount of work to implement (which was what was estimated for the JSOCINT ticket) whereas introducing the new filter offset would be significantly more work.    See [~bushouse] comment here  : https://jira.stsci.edu/browse/JWSTDMS-207

If our understanding or recommendation doesn't meet the need, then let's meet further to discuss it.

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: Thanks, [~scrawford]. I think [~jstans] wanted to continue to discuss the details of how this would work, I'm not sure that there was an agreement on exactly what changes needed to be made in the pipeline just yet. I was creating the placeholder tickets to get this on the radar. I think [~bushouse] was at the first meeting, and [~dencheva] participated in the discussion yesterday. Maybe we need another full discussion with SCSB folks and the INS WCS reference file reps?

**Comment by stscijgbot** [comment]
Comment by Steven Crawford: Also [~bushouse] reminded me that we do have [FILTEROFFSET reference files|https://jwst-pipeline.readthedocs.io/en/latest/jwst/references_general/filteroffset_reffile.html#filteroffset-reffile] already that are used in MIRI and so these are not new reference file types but rather new reference file for NIRCAM.

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: Yeah, we discussed whether or not we could use the existing FilteroffsetModel or whether we would need to update the datamodel(s) for different instruments. Nadia provided some options, but I think this is still TBD because not all instruments were present to comment.

**Comment by stscijgbot** [comment]
Comment by Alberto Noriega-Crespo: From the MIRI perspective, David Law spoke with Nadia and the correct units to describe the file have been introduced in the documentation.  Nadia is "tasked to change the file", but David can do it as well.

**Comment by stscijgbot** [comment]
Comment by David Law: I just created PR [https://github.com/spacetelescope/jwst/pull/4595] to add an input_units keyword to the filteroffset reference file schema.  Attached ([^jwst_miri_filteroffset_cdp7.asdf]) is the MIRI filter offset reference file created using this new schema; this can be delivered to crds after Nadia confirms that my PR is ok, and any iteration on any other changes to the schema.

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: The CIOWG recently discussed the need for filter offset reference files. MIRI currently has FILTEROFFSET reference files in CRDS (i.e., the datamodel exists already), which have this format (from [this link|https://jwst-pipeline.readthedocs.io/en/latest/jwst/references_general/filteroffset_reffile.html#filteroffset-reffile]):

The filteroffset reference file is an ASDF file that contains a list called {{filters}}. Every item in the list contains three fields - {{row_offset}}, {{column_offset}} and {{name}} - the name of the filter they are valid for. The offsets, in pixels, are applied in the image science frame.
  ||filters:|
  ||column_offset:|Offset in x (in pix)|
||row_offset:|Offset in y (in pix)|
||name:|Filter name|
|

------

NIRISS ([~kvolk] /[~goudfroo])
MIRI ([~dlaw] /[~cracraft]/[~anoriega])
NIRCam ([~jstans] /[~hilbert])
SCSB ([~dencheva] /[~bushouse])

To decide how we should proceed, can you chime in on the following?
 # If we update the FILTEROFFSET file to have a *filter,pupil* list instead of just filter, would that be a significant amount of work in terms of datamodel, code, and MIRI reference file updates? 
 # NIRCam is okay with the the current file format as long as we can add in the *pupil* list. NIRISS, is your team okay with the FILTEROFFSET format?

**Comment by stscijgbot** [comment]
Comment by David Law: What exactly is the proposed change?  To have an array called 'filters' that includes a column for both 'name' and 'pupil'?  To have a new array called 'pupil', or something else?

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: I believe we would just need to update the array and list names to allow the values to be selected based off of filter/pupil combinations. I think [~dencheva] said in the CIOWG meeting that it wouldn't be too much work to modify the datamodel to allow for this, but I'm not quite sure how that would affect the current MIRI files (i.e., could MIRI just keep the normal filter names and leave pupil blank?). Nadia, could you comment on the easiest way to update the datamodel to accommodate filter/pupil combinations needed by NIRCam and NIRISS, but not by MIRI? I'd rather not have to have separate datamodels, but if that's the only way to go, then we can head down that road, too.

Also, I realize that this work should be prioritized considering all the other tasks, I just want to get the ball rolling to see what need to be done.

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: I believe we would just need to update the array and list names to allow the values to be selected based off of filter/pupil combinations. I think [~dencheva] said in the CIOWG meeting that it wouldn't be too much work to modify the datamodel to allow for this, but I'm not quite sure how that would affect the current MIRI files (i.e., could MIRI just keep the normal filter names and leave pupil blank?). [~dencheva], could you comment on the easiest way to update the datamodel to accommodate filter/pupil combinations needed by NIRCam and NIRISS, but not by MIRI? I'd rather not have to have separate datamodels, but if that's the only way to go, then we can head down that road, too.

Also, I realize that this work should be prioritized considering all the other tasks, I just want to get the ball rolling to see what need to be done.

**Comment by stscijgbot** [comment]
Comment by Paul Goudfrooij: Just to add that indeed, NIRISS will also need the PUPIL column/parameter, since some of NIRISS' filters are in the pupil wheel and the relevant header keywords are PUPIL and FILTER.

**Comment by stscijgbot** [comment]
Comment by Kevin Volk: The general expectation for NIRISS is that we will make separate distortion files for each imaging filter, and that the filter offsets reference file is not needed.  We will not know whether the distortion is different for the different filters until we observe the astrometric field, but it seems likely to be the case.  In such an instance we would not be using the filter to filter offsets for the WCS step.  I would be rather surprised if we found that the 12 NIRISS filters all had the same distortion aside from filter to filter shifts (presumably from wedging effects in each specific filter).  The shifts have been measured in CV3, but not the distortion.

**Comment by stscijgbot** [comment]
Comment by Nadia Dencheva: The addition of "pupil" to the datamodels is made in [#4750|[https://github.com/spacetelescope/jwst/pull/4750]].

There's some discussion on different options in the PR and we converged on the following:

 
 * We keep the initial schema for MIRI imager so MIRI reference files will not change
 * The new schema adds a "pupil" field and renames the " name" field to "filter"

 

It may make sense for the future if MIRI starts writing their reference files using "filter" instead of "name" so that eventually we can remove the initial schema but that is not mandatory.

 

Here's an example of how to create reference file using "pupil" :

 
{code:java}
filters = [{'filter': 'F090W', 'pupil': 'GRISMR', 'row_offset': 1, 'column_offset': 1},
           {'filter': 'F070W', 'pupil': 'GRISMC', 'row_offset': 2, 'column_offset': 2}
          ]

with FilteroffsetModel(filters=filters, instrument='NIRCAM', strict_validation=True) as fo:
        fo.filters == filters
        fo.meta.description = "Filter offsets"
        fo.meta.reftype = "filteroffset"
        fo.meta.author = "Unknown"
        fo.meta.pedigree = "GROUND"
        fo.meta.useafter = "2019-12-01"
        fo.meta.instrument.detector = 'NRCA1'
        fo.validate()
        fo.save("nircam_filteroffset.asdf")

{code}

**Comment by stscijgbot** [comment]
Comment by Nadia Dencheva: The addition of "pupil" to the datamodels is made in [#4750|[https://github.com/spacetelescope/jwst/pull/4750]].

There's some discussion on different options in the PR and we converged on the following:

 
 * We keep the initial schema for MIRI imager so MIRI reference files will not change
 * The new schema adds a "pupil" field and renames the " name" field to "filter"

 

It may make sense for the future if MIRI starts writing their reference files using "filter" instead of "name" so that eventually we can remove the initial schema but that is not mandatory.

 

Here's an example of how to create reference file using "pupil" :

 
{code:java}
from jwst.datamodels import FilteroffsetModel    

filters = [{'filter': 'F090W', 'pupil': 'GRISMR', 'row_offset': 1, 'column_offset': 1},
           {'filter': 'F070W', 'pupil': 'GRISMC', 'row_offset': 2, 'column_offset': 2}
          ]

with FilteroffsetModel(filters=filters, instrument='NIRCAM', strict_validation=True) as fo:
        fo.filters == filters
        fo.meta.description = "Filter offsets"
        fo.meta.reftype = "filteroffset"
        fo.meta.author = "Unknown"
        fo.meta.pedigree = "GROUND"
        fo.meta.useafter = "2019-12-01"
        fo.meta.instrument.detector = 'NRCA1'
        fo.validate()
        fo.save("nircam_filteroffset.asdf")

{code}

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: [~dencheva]  would it be possible to use the bar convention for the detector values, e.g:


fo.meta.instrument.detector = 'NRCA1|NRCA2|NRCA3|NRCA4'

**Comment by stscijgbot** [comment]
Comment by Nadia Dencheva: [~acanipe] Is the goal to have one file per filter for all detectors? If so, IIRC the bar convention can be used with the "P" keywords to automatically create an rmap I added this line

[https://github.com/spacetelescope/jwst/pull/4750/files#diff-125b27859727fa6ba72652ccd60bfe5fR685]

 

However, if "detector" is not relevant it shouldn't be in the reference file. I wanted to provide as many default values as possible and facilitate validation so I added "detector" to the validation code. It sounds like I should remove it. [~bushouse] may correct me, I haven't dealt with reference files recently.

**Comment by stscijgbot** [comment]
Comment by Howard Bushouse: If these reference files are going to be detector-independent, then I don't think we need to have the detector keyword (or its parameterized version) in the headers of the reference files at all. In cases where keywords might help to provide pedigree-related info as to how the calibration data were obtained and created, then I can understand wanting to include them. But even then, those keywords probably shouldn't be listed in the model schema, because they aren't necessary for validation. So if the instrument teams indicate that these reference files will be detector-independent, then I'd say don't have a detector (or p_detector) keyword in them at all.

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: Thanks, [~bushouse] and [~dencheva]. The NIRCam team/[~jstans] was discussing how to build the files yesterday. I think the idea is that we need the flexibility to have one file for the module A shortwave detectors (with one table that includes all the relevant filters), one for module B shortwave, one for A longwave, and one for B longwave. So they would be channel/module dependent (but, for instance, all module A shortwave detectors would use the same values). John, does that describe it correctly?

**Comment by stscijgbot** [comment]
Comment by John Stansberry: Yes, Alicia's summary just above is correct re. the need to specify the filter offsets for each channel (SW, LW), and each module (A, B).

The filter + pupil combinations listed for the A & B SW channels will be the same, but the offsets for the two modules will differ. Similarly the A & B LW channels will have the same list of filter + pupil, but different offsets in the two modules.

**Comment by stscijgbot** [comment]
Comment by Howard Bushouse: So we don't need the detector keyword, but we do need CHANNEL ('SHORT', 'LONG') and MODULE ('A', 'B'). And there will be separate ref files for each of those 4 possible combinations, so we don't need the parameterized versions of those keywords (they'll only have a single value in each ref file). Does that sound right?

**Comment by stscijgbot** [comment]
Comment by Nadia Dencheva: The addition of "pupil" to the datamodels is made in [#4750|[https://github.com/spacetelescope/jwst/pull/4750]].

There's some discussion on different options in the PR and we converged on the following:

 
 * We keep the initial schema for MIRI imager so MIRI reference files will not change
 * The new schema adds a "pupil" field and renames the " name" field to "filter"

 

It may make sense for the future if MIRI starts writing their reference files using "filter" instead of "name" so that eventually we can remove the initial schema but that is not mandatory.

 

Here's an example of how to create reference file using "pupil" :

 
{code:java}
from jwst.datamodels import FilteroffsetModel    

filters = [{'filter': 'F090W', 'pupil': 'GRISMR', 'row_offset': 1, 'column_offset': 1},
           {'filter': 'F070W', 'pupil': 'GRISMC', 'row_offset': 2, 'column_offset': 2}
          ]

with FilteroffsetModel(filters=filters, instrument='NIRCAM', strict_validation=True) as fo:
        fo.filters == filters
        fo.meta.description = "Filter offsets"
        fo.meta.reftype = "filteroffset"
        fo.meta.author = "Unknown"
        fo.meta.pedigree = "GROUND"
        fo.meta.useafter = "2019-12-01"
        fo.meta.instrument.channel = "SHORT"
        fo.meta.instrument.module = "A"
        fo.validate()
        fo.save("nircam_filteroffset.asdf")

{code}

**Comment by stscijgbot** [comment]
Comment by Nadia Dencheva: [~bushouse] Yes, this sounds right to me. I've updated the example above and the schema in the PR.

**Comment by stscijgbot** [comment]
Comment by John Stansberry: Great! Sounds like this is ready to roll. I'll work on getting the files prepared in the next day or two.

Would it make sense to prepare one file and try that out, or just go for the set of 4?

Also, please:
 * Confirm that I should log delivery of those files on CRDS-346.
 * Clarify who I contact to "get sign-off from ReDCaT and SCSB/CRDS"?
 * My supposition is that after I create the files I will ask [~bbrooks] to do the actual submission to the test server - let me know if that's incorrect.

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: Yep, delivery should be captured in CRDS-346. We'll need to use the latest version of the jwst software to get the datamodel changes. If you'd rather just create one file (or all the files) or tables, then I can test it out and save it in the datamodel. Once we have a version of the files we are happy with, I can just upload them into CRDS-346 and then I can go through the delivery process with ReDCaT/CRDS folks.

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: Sorry, I just saw that CRDS-346 was created – the delivery should be tracked in CRDS-345 (346 is for the CRDS team to track updates they need to make to the system to prepare for the files). We'll need to use the latest version of the jwst software to get the datamodel changes. If you'd rather just create one file (or all the files) or tables, then I can test it out and save it in the datamodel. Once we have a version of the files we are happy with, I can just upload them into CRDS-345 and then I can go through the delivery process with ReDCaT/CRDS folks.

**Comment by stscijgbot** [comment]
Comment by Howard Bushouse: The datamodel update was just merged into the jwst repo via [https://github.com/spacetelescope/jwst/pull/4750], so it's available in the github master branch.

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: Great, thanks [~bushouse] and [~dencheva]. I'll grab the latest version.

**Comment by stscijgbot** [comment]
Comment by David Law: In an effort to keep things simple where possible I've revised the MIRI imaging filter offset reference file to use the 'filter' instead of 'name' keyword.  This file will require a change in the assign_wcs step in order to use the new keyword name.  I've filed PR https://github.com/spacetelescope/jwst/pull/4776 to make the necessary update to the pipeline (and remove the old schema), and CRDS ticket [https://jira.stsci.edu/browse/CRDS-347|https://jira.sts/] to deliver the new reference file.  Since the two interact I expect that there may be errors until both are integrated.

**Comment by stscijgbot** [comment]
Comment by David Law: In an effort to keep things simple where possible I've revised the MIRI imaging filter offset reference file to use the 'filter' instead of 'name' keyword.  This file will require a change in the assign_wcs step in order to use the new keyword name.  I've filed PR [https://github.com/spacetelescope/jwst/pull/4776] to make the necessary update to the pipeline (and remove the old schema), and CRDS ticket https://jira.stsci.edu/browse/CRDS-347 to deliver the new reference file.  Since the two interact I expect that there may be errors until both are integrated.

**Comment by stscijgbot** [comment]
Comment by Misty Cracraft: MIRI file has been delivered to the test server (CRDS-347). When code is updated, the file needs to be pushed to the ops server to be used with the code update.

**Comment by stscijgbot** [comment]
Comment by Catherine Kaleida: What's the status on this ticket?  It's assigned to [~acanipe], but it looks like this is waiting on the code to be updated.

**Comment by stscijgbot** [comment]
Comment by Howard Bushouse: All code updates have been made. JP-1291 JP-1292

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: Just waiting for confirmation from MIRI about the files, then I'll close this.

**Comment by stscijgbot** [comment]
Comment by Kevin Volk: I do not remember whether we made this type of file for NIRISS just in case it could be used or whether it was never made.  I have values for the filter to filter offsets and could make a file.

**Comment by stscijgbot** [comment]
Comment by David Law: As far as MIRI is concerned we're set; the updated file and the code to read it have both been live for a while now.

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: Hi [~kvolk], I had to refresh my memory on this. I think we assumed you all wouldn't want this file yet based on the comment above (https://jira.stsci.edu/browse/JP-1290?focusedCommentId=457526&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-457526), but let me know if that's not the case.

**Comment by stscijgbot** [comment]
Comment by Kevin Volk: Yes our original idea was that we do not want this at least not yet.  If in commissioning we see that the distortion is common between the filters aside from an offset we would then look into this option.

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: Sounds good, thanks [~kvolk]. I think we're all good here, so I'll close this.

**Comment by stscijgbot** [comment]
Comment by Alicia Canipe: Initial versions of the files are delivered and code is updated, so we will close this. Any problems or new files needed down the road will be tracked in new tickets.
